internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
 	okayWait = 1 * time.Second
 	killWait = 5 * time.Second
 	failWait = 10 * time.Second
+)
+
+const maxLogBytes = 100 * 1024
+
+// serviceState represents the state a service's state machine is in.
+type serviceState int
 
-	maxLogBytes = 100 * 1024
+const (
+	stateInitial serviceState = iota
+	stateStarting
+	stateRunning
+	stateTerminating
+	stateKilling
+	stateStopped
+	stateBackoffWait
 )
 
-// Start starts the named service after also starting all of its dependencies.
+func (s serviceState) String() string {
+	switch s {
+	case stateInitial:
+		return "initial"
+	case stateStarting:
+		return "starting"
+	case stateRunning:
+		return "running"
+	case stateBackoffWait:
+		return "backoff-wait"
+	case stateTerminating:
+		return "terminating"
+	case stateKilling:
+		return "killing"
+	case stateStopped:
+		return "stopped"
+	default:
+		return "unknown"
+	}
+}
+
+// service holds the information for a service under our control.
+type service struct {
+	lock         sync.Mutex
+	manager      *ServiceManager
+	state        serviceState
+	config       *plan.Service
+	logs         *servicelog.RingBuffer
+	started      chan *int
+	stopped      chan error
+	cmd          *exec.Cmd
+	backoffIndex int
+	resetTimer   *time.Timer
+}
+
 func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
+	m.state.Lock()
+	request, err := TaskServiceRequest(task)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get service request: %w", err)
+	}
+
 	releasePlan, err := m.acquirePlan()
 	if err != nil {
+		return fmt.Errorf("cannot acquire plan lock: %w", err)
+	}
+	config, ok := m.plan.Services[request.Name]
+	if !ok {
+		releasePlan()
+		return fmt.Errorf("cannot find service %q in plan", request.Name)
+	}
+	releasePlan()
+
+	m.servicesLock.Lock()
+	s := m.services[config.Name]
+	if s != nil && s.getStateLocked() != stateStopped {
+		m.servicesLock.Unlock()
+		m.state.Lock()
+		task.Logf("Service %q already started.", config.Name)
+		m.state.Unlock()
```

Comment:
Yeah, you weren't wrong, it was a bit crazy. I've made most of these changes, and it really simplifies doStart/doStop. In particular:

1) Unified: gotten rid of the individual `serviceData.lock` and just use the "global" `servicesLock` everywhere. As you say, this locking/unlocking shouldn't be happening a ton, so it's not perf-critical.
2) Added helper methods to get the service, moving the `servicesLock` lock/unlock out of doStart/doStop and into those methods, where it can be a simple lock/defer unlock at the top.
3) Moved the lock-task.Logf-unlock dance to a helper function to clean that up.